### PR TITLE
Common functionality in Mesh type IObjects

### PR DESCRIPTION
**Pull request by OwenArnold**
`MeshObject` and `MeshObject2D` have the potential for more common/shared code. Remove duplication and adapt both into a form where code is cleaner. Should reduce the potential for misunderstanding and aid maintenance in the future.

**Tester**

* Code review is most important
* Instrument View should also be checked
* Performance tests should be checked post merge

### CODE REVIEW

**Review by nvaytet — CHANGES_REQUESTED** [review]
Seems all very good to me.
Some minor changes including possibly using this opportunity to clean up some namespace specifications that are not always needed/consistent.
In addition, maybe the
```
void MeshObject::getBoundingBox(double &xmax, double &ymax, double &zmax,
                                double &xmin, double &ymin,
                                double &zmin)
```
function can also be moved into the common file.

File: Framework/Geometry/src/Objects/MeshObjectCommon.cpp [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  // rest is just to solve for u, v, t and check u and v are both >= 0 and <= 1
+  // and u+v <=1
+
+  auto edge1 = v2 - v1;
+  auto edge2 = v3 - v1;
+  auto h = direction.cross_prod(edge2);
+  auto a = edge1.scalar_prod(h);
+
+  const double EPSILON = 0.0000001 * edge1.norm();
+  if (a > -EPSILON && a < EPSILON)
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/Geometry/src/Objects/MeshObjectCommon.cpp]: I didn't write that, but I suspect that's the case.

**Review by nvaytet — DISMISSED** [review]

**Comment by SimonHeybrock** [comment]
@OwenArnold This has been approved, but I see quite a few un-adressed comments. Is this ready or not?

File: Framework/Geometry/src/Objects/MeshObject.cpp [hunk 2]
```diff
@@ -605,13 +466,7 @@ size_t MeshObject::numberOfTriangles() const { return m_triangles.size() / 3; }
...
  */
 std::vector<uint32_t> MeshObject::getTriangles() const {
-  std::vector<uint32_t> faces;
-  size_t nFaceCorners = m_triangles.size();
-  faces.resize(static_cast<std::size_t>(nFaceCorners));
-  for (size_t i = 0; i < nFaceCorners; ++i) {
-    faces[i] = static_cast<int>(m_triangles[i]);
-  }
-  return faces;
+  return MeshObjectCommon::getTriangles_uint32(m_triangles);
```
[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/Geometry/src/Objects/MeshObject.cpp]: Yes exactly

**Comment by OwenArnold** [comment]
@nvaytet I've made changes but will require a re-review.

**Review by nvaytet — CHANGES_REQUESTED** [review]
Just a couple of comments, on the solid angle tests mainly

**Review by nvaytet — DISMISSED** [review]
Good to go.

**Comment by SimonHeybrock** [comment]
THere is a merge conflict now.

File: Framework/Geometry/CMakeLists.txt [hunk 3]
```diff
@@ -106,8 +106,8 @@ set ( SRC_FILES
 	src/Objects/BoundingBox.cpp
 	src/Objects/CSGObject.cpp
 	src/Objects/InstrumentRayTracer.cpp
+        src/Objects/MeshObject2D.cpp
```
Comment (soininen) [inline Framework/Geometry/CMakeLists.txt]: Intentation.

**Review by soininen — CHANGES_REQUESTED** [review]

**Review by soininen — DISMISSED** [review]

**Comment by OwenArnold** [comment]
@mantidproject/gatekeepers could someone please merge this it is holding up #23873

File: Framework/Geometry/src/Objects/MeshObject2D.cpp [hunk 4]
```diff
@@ -6,6 +6,7 @@
 // SPDX - License - Identifier: GPL - 3.0 +
 #include "MantidGeometry/Objects/MeshObject2D.h"
 #include "MantidGeometry/Objects/IObject.h"
+#include "MantidGeometry/Objects/MeshObjectCommon.h"
```
Comment (gvardany) [inline Framework/Geometry/src/Objects/MeshObject2D.cpp]: Includes are out of order here

File: Framework/Geometry/src/Objects/MeshObjectCommon.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
+#include "MantidGeometry/Objects/MeshObjectCommon.h"
+#include "MantidGeometry/Objects/BoundingBox.h"
```
Comment (gvardany) [inline Framework/Geometry/src/Objects/MeshObjectCommon.cpp]: and here

File: Framework/Geometry/src/Objects/MeshObject2D.cpp (hunk 4 above)
Comment (OwenArnold) [reply Framework/Geometry/src/Objects/MeshObject2D.cpp]: I agree here.

File: Framework/Geometry/src/Objects/MeshObjectCommon.cpp (hunk 5 above)
Comment (OwenArnold) [reply Framework/Geometry/src/Objects/MeshObjectCommon.cpp]: I think this should not change as there needs to be exceptions for implementation files of components including their definitions ["A component's .cpp file includes its corresponding .h file as the first substantive line of code"](https://github.com/bloomberg/bde/wiki/physical-code-organization)

John Lakos goes into detail on this. in his "Large Scale C++ Design" book. I'll change it anyway, but maybe this is something for the TSC to discuss.

Comment (gvardany) [reply Framework/Geometry/src/Objects/MeshObjectCommon.cpp]: Correct. Sorry for the noise.

File: Framework/Geometry/src/Objects/MeshObject2D.cpp (hunk 4 above)
Comment (gvardany) [reply Framework/Geometry/src/Objects/MeshObject2D.cpp]: It's correct here as well then.

**Comment by gvardany** [comment]
Please revert the last commit, and I'll merge.

File: Framework/Geometry/src/Objects/MeshObject2D.cpp (hunk 4 above)
Comment (martyngigg) [reply Framework/Geometry/src/Objects/MeshObject2D.cpp]: clang-format should sort out the include order these days. It has an exception that keeps the header for the current implementation file at the top.

**Review by soininen — APPROVED** [review]

**Pull request merged**